### ISSUE TITLE
Teacher Tools: Setting the time to spend in each topic

### ISSUE THREAD

**Issue by IvonArroyo** [center-for-knowledge-communication/mathspring#147]
We need to allow teachers to set properties of the behavior of topics, particularly the timing and the number of problems to see in each topic.

This involves changing the default time for each student to spend in each topic, so that students don't get "stuck" in a single topic. 

At the same time, if a teacher has only enabled a single topic, we don't want the student to be invited to get out of the topic during a full class period.

**Comment by IvonArroyo** [comment]
More details about this: 

The ClassConfig table has configurations for each class tthat allow you to regulate this. 

`**maxNumberProbsToShowPerTopic**` specifies the maximum number of problems to show in a topic before the student is moved on to the next topic.
`**maxTimeInTopic**` specifies the maximum time that a student will spend in a topic before the student is moved on to the next topic.

`**minNumberProbsToShowPerTopic**` specifies the minimim number of problems the student has to see in a topic before the student is moved on to the next topic.
`**minTimeInTopic**` specifiies the minimum time tthe student has to spend in a topic before the student is moved on to the next topic.

MOVING ON to the next topic means that, when the Java code requests the topic somewhere, it will return a new topic number, the next one. Somehow the code manages to show a Message in a window before switching to the next topic.

May I suggest that you tweak these values in a class, and see what happens when you log iin as a user to that class. Set low maximums so that you don't have to wait much.

We should set the maximum time in a topic every time the teacher updates the active topics through the problem manager. Assuming a student will spend 45 minutes in a day after they logged in,  If a teacher enables 1 topic, the maximum time should be 45 minutes.

If a teacher enables 2 topics, then the maximum time in a topic should be 23 minutes. 
If a teacher enables 3 topics the maximum time should be 15 minutes.
If a teacher enables 4 topics, the max time should be 11 minutes.
In general, divide 45 minutes by the number of topics.
HOWEVER, the max time should never be less than 10 minutest. 
So, If a teacher enables 5 topics or more,  the maximum time should be 10 minutes. 

We could leave everything as is for all the other ones (max problems, min problems, min time) for the time being.  

We could also change the message at the end of the topic, which pops up for the end of topic condition should be adjusted. 

I can do this I guess.

Java Classes regarding this are:

package edu.umass.ckc.wo.tutor.pedModel.EndOfTopicInfo
package. edu.umass.ckc.wo.tutor.pedModel.TopicSelectorImpl. (see public EndOfTopicInfo isEndOfTopic())

**Comment by mkartik** [comment]
The current message when the maximum time in a topic is reached is:
_You are about to switch topics because:
You've spent a lot of time in this topic.
Let's try something else!

You have solved 10 / 10 of the problems you saw._

Should this message be updated, as part of this issue?

**Comment by perspegrity5** [comment]
As Kartik reported on finishing implementation I found myself wondering if this strategy would always work.  Some topics cover several standards and dozens of problems, so some students may end up spending a lot of time on them. If a teacher enables 4 topics, that is for her *class*, not for one class session, right? It could be over many classes, so does it make sense to assume 45 minutes? I think it may need more discussion.

**Comment by perspegrity5** [comment]
Also, when a student times out on a topic without mastery, don't we circle them back around later? Is there a message for that?

**Comment by IvonArroyo** [comment]
No, it is for every session, not for every class. Every time a student logs in, all these timers gets reset. 
So this is a good strategy for the time being, we really need something so that students don't just get stuck in the first topic all the time.

--Ivon

**Comment by IvonArroyo** [comment]
You are about to switch topics because:
You've spent a lot of time in this topic.
Let's try something else! You can also go to "My Progress" or to see how you are doing, or go to "My Garden" and see how much your plants are growing. 

You have solved 10 / 10 of the problems you saw._

**Comment by IvonArroyo** [comment]
This might be a good time to move to a new topic because [you have spent a lot of time on this topic / or other reason]. Click on of these do you want to do:  <Go to the next topic: [Fractions]> <Stay on this topic longer> <Visit the Progress Page and Choose the next topic from there>

**Comment by IvonArroyo** [comment]
This would imply resetting a timer if they choose to "stay on this topic longer".
